app-modules/engagement/src/Filament/Actions/SendEngagementAction.php

```diff
@@ -184,6 +256,15 @@ protected function setUp(): void
                                         NotificationChannel::Sms => $educatable?->phoneNumbers()
                                             ->where('can_receive_sms', true)
                                             ->get()
+                                            ->filter(function (StudentPhoneNumber | ProspectPhoneNumber $phoneNumber) {
```

Comment:
Instead of filtering after the phone numbers have been fetched, you could use a `whereDoesntHave()` to scope the results to only those that aren't related to an opt out record.

The opt out records could be associated with a phone number using a `BelongsTo` Eloquent relationship, using `number` as the foreign key.